astropy_bot/autolabel.py

```diff
@@ -0,0 +1,24 @@
+from os import path, sep
+from baldrick.plugins.github_pull_requests import pull_request_handler
+
+
+@pull_request_handler
+def autolabel_subpackages(pr_handler, repo_handler):
+    files = pr_handler.get_modified_files()
+    all_labels = repo_handler.get_all_labels()
+    existing_labels = set(pr_handler.labels)
+
+    for file in files:
+        subdir = path.dirname(file)
+
+        if subdir:
+            root, *subpkg = subdir.split(sep)
+
+            if subpkg:
+                dot_name = '.'.join(subpkg)
+                if dot_name in all_labels and dot_name not in existing_labels:
```

Comment:
What if only `astropy/io/fits/tests/test_something.py` is modified? Then this won't work because it would be looking for `io.fits.tests`?

What about `astropy/tests/something.py`?

What about `docs/units/something.rst`?

What about `astropy_helpers` submodule update?

What about `setup.py` or `.circleci` in the root directory?

I am not saying you have to support all that. I am just listing them for consideration. Also, when customization is added, can also check if it is something to be skipped before labeling.

Reply:
* [ ] astropy/io/fits/tests/test_something.py
* [ ] docs/units/something.rst

What labels do these cases map to, and do we want to set these as well?
* What about astropy/tests/something.py?
* What about astropy_helpers submodule update?
* What about setup.py or .circleci in the root directory?

Reply:
In *general*, at least, when I label stuff, it goes more or less like this:

* `astropy/io/fits/tests/test_something.py` (and only that file) -- `io.fits` + `testing` + `no-changelog-entry-needed`
* `astropy/io/fits/tests/test_something.py` and `astropy/io/fits/code.py` -- `io.fits`
* `docs/units/something.rst` (and only that file) -- `units` + `Docs` + `no-changelog-entry-needed`
* `docs/units/something.rst` and `astropy/units/code.py` -- `units`
* `astropy_helpers` (with or without something else) -- `astropy-helpers`
* `.circleci`, `.travis.yml`, or `appveyor.yml` -- `testing` + `no-changelog-entry-needed`
* `setup.py` or `setup.cfg` -- `installation`

Is this enough? 😅  Of course, this is not exhaustive, but we can add more as needed in the future.

Reply:
I think the rest of this logic should be done in a separate PR, probably by you since you are most familiar with the flow (maybe that should be in astropy/pllim-bot??)! I'm only going to touch auto-labeling subpackages here, per the issue and PR scope. But I left room (and a TODO comment) where you can add more.

Reply:
Sounds good.